pkg/controllers/managementuser/rbac/roletemplates/roletemplate_handler.go

```diff
@@ -0,0 +1,232 @@
+	"storageclasses":    "storage.k8s.io",
+	"apiservices":       "apiregistration.k8s.io",
+	"clusterrepos":      "catalog.cattle.io",
+	"clusters":          "management.cattle.io",
+}
+
+const (
+	clusterRoleOwner = "authz.cluster.cattle.io/clusterrole-owner"
+)
+
+func newRoleTemplateLifecycle(uc *config.UserContext) *roleTemplateLifecycle {
+	return &roleTemplateLifecycle{
+		crClient: uc.RBAC.ClusterRoles(""),
+	}
+}
+
+type roleTemplateLifecycle struct {
+	crClient typesrbacv1.ClusterRoleInterface
+}
+
+// Create func always creates:
+// 1. a ClusterRole with the same name as the RoleTemplate
+// 2. an Aggregating ClusterRole that aggregates all inherited RoleTemplates with the name "RoleTemplateName-aggregator"
+//
+// For RoleTemplates with the Context == "Project", we also create:
+// 1. If the RoleTemplate has any rules for Global Resources, make a ClusterRole with those named "RoleTemplateName-promoted"
+// 2. an Aggregating ClusterRole that aggregates all inherited RoleTemplates' promoted Cluster Roles named "RoleTemplateName-promoted-aggregator"
+func (rtl *roleTemplateLifecycle) Create(rt *v3.RoleTemplate) (runtime.Object, error) {
+	// 1. Create cluster role
+	cr := rtl.buildClusterRole(rt)
+	_, err := rtl.crClient.Create(cr)
+	if err != nil {
+		return nil, err
+	}
+	// 2. Create aggregating cluster role
+	cr = rtl.buildAggregatingClusterRole(rt)
+	_, err = rtl.crClient.Create(cr)
+	if err != nil {
+		return nil, err
+	}
+	// 3. Project global resources
+	if rt.Context == "project" {
+		// Create a promoted cluster role
+		cr = rtl.buildProjectPromotedClusterRole(rt)
+		_, err = rtl.crClient.Create(cr)
+		if err != nil {
+			return nil, err
+		}
+
+		// Create a promoted aggregating cluster role
+		cr = rtl.buildProjectPromotedAggregatedClusterRole(rt)
+		_, err = rtl.crClient.Create(cr)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return rt, nil
+}
+
+func (rtl *roleTemplateLifecycle) Updated(rt *v3.RoleTemplate) (runtime.Object, error) {
+	// 1. Create cluster role
+	cr := rtl.buildClusterRole(rt)
+	_, err := rtl.crClient.Update(cr)
+	if err != nil {
+		return nil, err
+	}
+	// 2. Create aggregating cluster role
+	cr = rtl.buildAggregatingClusterRole(rt)
+	_, err = rtl.crClient.Update(cr)
+	if err != nil {
+		return nil, err
+	}
+	// 3. Project global resources
+	if rt.Context == "project" {
+		// Create a promoted cluster role
+		cr = rtl.buildProjectPromotedClusterRole(rt)
+		_, err = rtl.crClient.Update(cr)
+		if err != nil {
+			return nil, err
+		}
+
+		// Create a promoted aggregating cluster role
+		cr = rtl.buildProjectPromotedAggregatedClusterRole(rt)
+		_, err = rtl.crClient.Update(cr)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return rt, nil
+}
+
+func (rtl *roleTemplateLifecycle) Remove(rt *v3.RoleTemplate) (runtime.Object, error) {
+	err := errors.Join(
+		rtl.crClient.Delete(rt.Name, &metav1.DeleteOptions{}),
+		rtl.crClient.Delete(rt.Name+"-aggregator", &metav1.DeleteOptions{}))
```

Comment:
Please use functions for these. E.g.
```
func clusterRoleNameFor(rt *v3.RoleTemplate) string {
	return rt.Name
}


func aggregatorClusterRoleNameFor(rt *v3.RoleTemplate) string {
	return rt.Name+"-name"
}

...
rtl.crClient.Delete(clusterRoleNameFor(rt), &metav1.DeleteOptions{}),
```

Reply:
You also need to be conscious of retries: deleting an object that doesn't exist will return an error, so you need to ignore those errors too

Reply:
Fixed so it checks for "IsNotFound". For delete I prefer to attempt to delete everything even if there is an error for some of the delete operations, which is why I use `errors.Join` here